src/theory/strings/regexp_elim.cpp

```diff
@@ -0,0 +1,441 @@
+    return returnElim(atom, ret, "concat-splice");
+  }
+  Assert(nchildren > 1);
+  for (unsigned i = 0; i < nchildren; i++)
+  {
+    if (children[i].getKind() == STRING_TO_REGEXP)
+    {
+      Node s = children[i][0];
+      Node lens = nm->mkNode(STRING_LENGTH, s);
+      // there exists an index in this string such that the substring is this
+      Node k;
+      std::vector<Node> echildren;
+      if (i == 0)
+      {
+        k = d_zero;
+      }
+      else if (i + 1 == nchildren)
+      {
+        k = nm->mkNode(MINUS, lenx, lens);
+      }
+      else
+      {
+        k = nm->mkBoundVar(nm->integerType());
+        Node bound =
+            nm->mkNode(AND,
+                       nm->mkNode(LEQ, d_zero, k),
+                       nm->mkNode(LT, k, nm->mkNode(MINUS, lenx, lens)));
+        echildren.push_back(bound);
+      }
+      Node substrEq = nm->mkNode(STRING_SUBSTR, x, k, lens).eqNode(s);
+      echildren.push_back(substrEq);
+      if (i > 0)
+      {
+        std::vector<Node> rprefix;
+        rprefix.insert(rprefix.end(), children.begin(), children.begin() + i);
+        Node rpn = TheoryStringsRewriter::mkConcat(REGEXP_CONCAT, rprefix);
+        Node substrPrefix = nm->mkNode(
+            STRING_IN_REGEXP, nm->mkNode(STRING_SUBSTR, x, d_zero, k), rpn);
+        echildren.push_back(substrPrefix);
+      }
+      if (i + 1 < nchildren)
+      {
+        std::vector<Node> rsuffix;
+        rsuffix.insert(rsuffix.end(), children.begin() + i + 1, children.end());
+        Node rps = TheoryStringsRewriter::mkConcat(REGEXP_CONCAT, rsuffix);
+        Node ks = nm->mkNode(PLUS, k, lens);
+        Node substrSuffix = nm->mkNode(
+            STRING_IN_REGEXP,
+            nm->mkNode(STRING_SUBSTR, x, ks, nm->mkNode(MINUS, lenx, ks)),
+            rps);
+        echildren.push_back(substrSuffix);
+      }
+      Node body = nm->mkNode(AND, echildren);
+      if (!k.isNull())
+      {
+        Node bvl = nm->mkNode(BOUND_VAR_LIST, k);
+        body = nm->mkNode(EXISTS, bvl, body);
+      }
+      // e.g. x in re.++( R1, "AB", R2 ) --->
+      //  exists k.
+      //    0 <= k <= (len(x)-2) ^
+      //    substr( x, k, 2 ) = "AB" ^
+      //    substr( x, 0, k ) in R1 ^
+      //    substr( x, k+2, len(x)-(k+2) ) in R2
+      return returnElim(atom, body, "concat-find");
+    }
+  }
+  return Node::null();
+}
+
+Node RegExpElimination::eliminateStar(Node atom)
+{
+  if (!options::regExpElimAgg())
+  {
+    return Node::null();
+  }
+  // only aggressive rewrites below here
+
+  NodeManager* nm = NodeManager::currentNM();
+  Node x = atom[0];
+  Node lenx = nm->mkNode(STRING_LENGTH, x);
+  Node re = atom[1];
+  // for regular expression star,
+  // if the period is a fixed constant, we can turn it into a bounded
+  // quantifier
+  std::vector<Node> disj;
+  if (re[0].getKind() == REGEXP_UNION)
+  {
+    for (const Node& r : re[0])
+    {
+      disj.push_back(r);
+    }
+  }
+  else
+  {
+    disj.push_back(re[0]);
+  }
```

Comment:
I'm not sure I understand this syntax. How does this work if re is not kind `REGEXP_UNION`?

Reply:
This is a [kinded_iterator](https://github.com/CVC4/CVC4/blob/master/src/expr/node.h#L721-L732). It iterates over the children of the node if the node matches the kind and otherwise just iterates over the node itself.

Reply:
Interesting. I think I'd prefer keeping this code since perhaps there would be other ways to infer a regular expression is disjunctive.